### PR TITLE
Move remaining tally filters to C++

### PR DESCRIPTION
**Pull request by smharper**
i revisited the filters like `EnergyoutFilter` and `DelayedGroupFilter` that we abuse a bit in tally.F90, and I found that it wasn't actually that hard to move them over to C++.  With this PR, those filters and all the rest now have most of their implementation in C++.  This allowed a bit of clean-up like removing the `CppTallyFilter` class.  I also cut out one of the wrapper function calls for `get_all_bins` that was in place just for easy debugging while converting the filters.

Here I've also decided to start using `unique_ptr<Filter>` in place of `Filter*`.  I was originally holding back on that because the pointers aren't really unique since Fortran carries copies around, but I think the resource safety provided by `unique_ptr` is worth that little bit of broken contract.

### CODE REVIEW

**Comment by pshriwise** [comment]
@smharper I noticed that travis failed here due to some issues with the DAGMC configuration for those builds. Wanted to let you know I'm testing a fix for that and will submit it soon.

File: src/tallies/filter_particle.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{
+  return "Particle " + std::to_string(particles_[bin-1]);
+}
+
+//==============================================================================
+// Fortran interoperability
+//==============================================================================
+
+extern "C" int particle_filter_particles(ParticleFilter* filt, int i)
+{return filt->particles_[i];}
```
[the remark this answers is not in the document]
Comment (smharper) [reply src/tallies/filter_particle.cpp]: It should be.  Good catch!

File: src/tallies/filter_energy.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+#include "openmc/xml_interface.h"
+
+namespace openmc {
+
+//==============================================================================
+// EnergyFilter implementation
+//==============================================================================
+
+// Used to grab the rev_energy_bins array defined in Fortran.
+extern "C" double* rev_energy_bins_ptr();
```
[the remark this answers is not in the document]
Comment (smharper) [reply src/tallies/filter_energy.cpp]: So we do.  It looks like it's reversed from what we have in Fortran though.  I've added a commit that makes them match now.

**Comment by paulromano** [comment]
Thanks for catching the `rev_energy_bins` ordering. I now see the logic is a little confusing on the Fortran side -- not sure why the data from the file isn't read directly into `rev_energy_bins` there.

**Pull request merged**